docs/deployment.rst

```diff
@@ -27,6 +27,24 @@ The host or Docker image must be set up with a sandbox user, sandbox virtualenv,
 
 See `2U's Dockerfile <https://github.com/edx/public-dockerfiles/blob/main/dockerfiles/codejail-service.Dockerfile>`__ for an example of how you could set up your host or container. (As of March 2025, there is `not yet a Tutor plugin <https://github.com/openedx/codejail-service/issues/26>`__.)
 
+App user UID
+============
+
+The UID under which the webapp runs must be running as few other processes as possible, otherwise the ``NPROC`` setting will need to be elevated to the point of uselessness. This is due to an interaction of several kernel features:
+
+* The ``RLIMIT_NPROC`` mechanism that we use to limit the number of processes the sandbox process can create actually shares its usage pool with *all* of the processes owned by the same UID. If a process has ``NPROC`` set and needs to fork, that value must be higher than the *existing* number of processes (and threads) owned by the same UID.
+* Docker cannot isolate UIDs in containers, just the *mapping* of usernames to UIDs. If the web service runs as ``app`` inside the container but has user ID 1000 (the most likely situation), the kernel doesn't see any distinction between that and a user on the host that also has UID 1000.
+
+If the webapp runs with UID 1000, and there is also a user with UID 1000 on the host, then they participate in the same usage pool. On a typical Linux desktop, it is common for there to be ~1200 processes and threads already in use by the default user. Running codejail-service in a Docker container on that host with UID 1000 would require ``CODE_JAIL.limits.NPROC`` to be set to 2000 or so in order to ensure there is headroom. This is a much higher limit than we actually want for a sandbox.
+
+On an actual deployment host with no regular users, just system users (with UID < 1000), and no other services running in other containers, this might work. However, the most likely deployment situation is in Kubernetes, where there may be a number of other pods; many of these pods will be using UID 1000. But even on a traditional host it creates a fragile situation, where starting unrelated processes as UID 1000 would cause codejail to mysteriously start breaking.
+
+One solution is to select large, random values for the app and sandbox users. On a modern Linux kernel, UIDs and GIDs can be as high as ``2^32 - 1``, although from testing it may be safer to constrain to ``2^31 - 1``. Recommendation:
+
+* Select a UID and GID for the webapp user: ``python3 -c 'import random; print(random.randrange(3000, 2 ** 31))'``. Create a group and user for the webapp user using this as its UID and GID.
+* Do the same for the sandbox user, with a new value.
+* Restrict the number of codejail-service instances that are permitted to be scheduled on the same host as each other.
```

Comment:
1. I'd break out the section starting with `Recommendation: ...` and move it to the top. You may need to adjust the language a bit, but you can ask the reader to continue reading for background to understand the recommendation.
2. Do we want to propose a value (or range for ``RLIMIT_NPROC``) that completes this?
3. I'm not certain I know what the line `Restrict the number of codejail-service instances...` means, and how it relates to the different settings or other config.

Reply:
1. Sounds good, will do.
2. It might be better to leave NPROC recommendations to the codejail library docs, in this case. I'm planning on adding some text there about limitations of that feature, and that's probably a good place to say something about how to establish an appropriate minimum.
3. That line is really about Kubernetes (or other orchestration) configs, rather than the host config itself.  The problem is that if you have multiple codejail pods scheduled on the same host, they'll still have the same UID and will share a process usage pool. So scheduling them onto different hosts could be helpful.

Reply:
Made some adjustments, see what you think.